client/connection.go

```diff
@@ -440,14 +449,51 @@ func (conn *Conn) recv() {
 // ping is started as a goroutine after a connection is established, as
 // long as Config.PingFreq >0. It pings the server every PingFreq seconds.
 func (conn *Conn) ping() {
-	defer conn.wg.Done()
 	tick := time.NewTicker(conn.cfg.PingFreq)
+	var sentPings []string
+
 	for {
 		select {
 		case <-tick.C:
-			conn.Ping(fmt.Sprintf("%d", time.Now().UnixNano()))
+			logging.Debug("irc.ping(): Current outstanding pings: %d/%d",
+				len(sentPings), conn.cfg.FailedPings)
+			// If we haven't cleared this by now it's time to bail.
+			if conn.cfg.FailedPings > 0 && len(sentPings) >= conn.cfg.FailedPings {
+				logging.Error("irc.ping(): Failed to recieve %d pings, closing down the connection.",
+					conn.cfg.FailedPings)
+				tick.Stop()
+				conn.wg.Done()
+				conn.Close()
+				return
+			}
+
+			currentPing := fmt.Sprintf("%d", time.Now().UnixNano())
+			sentPings = append(sentPings, currentPing)
+
+			conn.Ping(currentPing)
+		case pong := <-conn.pongResponses:
+			// No need to waste the cycles if we aren't tracking the number of failed pings.
+			if conn.cfg.FailedPings == 0 {
+				continue
+			}
+
+			logging.Debug("irc.ping(): Processing Pong: %s", pong)
+
+			// See if ping matches
+			shouldClear := false
+			for _, sentPing := range sentPings {
+				if strings.Contains(pong, sentPing) {
```

Comment:
For this particular piece I'm not sure what to do. At least on Freenode / the other places I've tested it will work fine if you just do `pong == sentPing`. I think contains is probably the safer bet though.